Sources/SwiftDocC/Infrastructure/Link Resolution/PathHierarchy.swift

```diff
@@ -148,7 +148,48 @@ struct PathHierarchy {
                 // would require that we redundantly create multiple nodes for the same symbol in many common cases and then merge them. To avoid doing that, we instead check
                 // the source symbol's path components to find the correct target symbol by matching its name.
                 if let targetNode = nodes[relationship.target], targetNode.name == expectedContainerName {
-                    targetNode.add(symbolChild: sourceNode)
+                    if sourceNode.parent == nil || sourceNode.parent === targetNode {
+                        targetNode.add(symbolChild: sourceNode)
+                    } else {
+                        // If the node we have for the child has an existing parent that doesn't
+                        // match the parent from this symbol graph, we need to clone the child to
+                        // ensure that the hierarchy remains consistent.
+                        let clonedSourceNode = Node(cloning: sourceNode, children: [:])
+
+                        // Because we're creating a new node to represent this language's counterpart symbol,
+                        // clean up the languages sets for the original and clone node.
+                        sourceNode.languages.remove(language!)
+                        clonedSourceNode.languages = .init([language!])
+
+                        // Also make sure that the symbol data is updated with this graph's symbol.
+                        clonedSourceNode.symbol = graph.symbols[relationship.source]
+
+                        // Make sure that the clone's children can all line up with symbols from this symbol graph.
+                        for (childName, children) in sourceNode.children {
+                            for child in children.storage {
+                                guard let childSymbol = child.node.symbol else {
+                                    // We shouldn't come across any non-symbol nodes here,
+                                    // but assume they can work as child of both variants.
+                                    clonedSourceNode.add(child: child.node, kind: child.kind, hash: child.hash)
+                                    continue
+                                }
+                                if nodes[childSymbol.identifier.precise] === child.node {
+                                    clonedSourceNode.add(symbolChild: child.node)
+                                }
+                            }
+                        }
```

Comment:
This block was added in a fit of paranoia, since i already needed to update the symbol data. I'm not sure if this is correct, but if C unions are the only thing that trigger this scenario this code shouldn't be hit in the first place. (Nested types in C create their declarations at top-level, so even declaring a type inline with a name won't create grandchild nodes of the union itself.)